### PR TITLE
Run API-calling validators only when related attribute is changed

### PR DESCRIPTION
**Pull request by EricPickup**
## Problem
We have validators such as `starter_code_repository_not_empty` and `starter_code_repository_is_template` that need to call the API to run properly. These validators are run every single time an assignment is created or updated (regardless of whether the starter code repo is changed or added).

## Solution
Run these validators only when `starter_code_repo_id` is added or changed.

I ran a few before/after benchmarks for a simple update to an `Assignment` and it went from `800ms` to `30ms`. 😄 

Side note: Moved `starter_code_repository_not_empty` into the `StarterCodeImportable` concern to get rid of duplicate methods in `Assignment`/`GroupAssignment`.

### CODE REVIEW

**Comment by shaunakpp** [comment]
Wow! Nice catch! 👏

**Comment by EricPickup** [comment]
@shaunakpp @d12 caught it actually 👏

**Comment by shaunakpp** [comment]
We should evaluate other such API calls across the app too...

File: app/models/assignment.rb [hunk 1]
```diff
@@ -38,9 +38,8 @@ class Assignment < ApplicationRecord
   validates :assignment_invitation, presence: true
 
   validate :uniqueness_of_slug_across_organization
-  validate :starter_code_repository_not_empty
-
-  validate :starter_code_repository_is_template
+  validate :starter_code_repository_not_empty, if: :will_save_change_to_starter_code_repo_id?
```
Comment (d12) [inline app/models/assignment.rb]: Wow, til these `will_save_change_to_starter_code_repo_id?` methods. Nice find :sparkles:

Comment (shaunakpp) [reply app/models/assignment.rb]: Rails magic ✨

**Pull request merged**